db/attachment_test.go

```diff
@@ -1359,3 +1359,129 @@ func TestMigrateBodyAttachmentsMergeConflicting(t *testing.T) {
 	bodyAtts, foundBodyAtts := body1[BodyAttachments]
 	assert.False(t, foundBodyAtts, "not expecting '_attachments' in body but found them: %v", bodyAtts)
 }
+
+func TestAllowedAttachments(t *testing.T) {
+	defer base.SetUpTestLogging(base.LevelDebug, base.KeySync)()
+
+	var tests = []struct {
+		name              string
+		inputBlipProtocol string
+		inputAttVersion   int
+	}{
+		{"TestAllowedAttachmentsCBMobile2AttVer1", BlipCBMobileReplicationV2, AttVersion1},
+		{"TestAllowedAttachmentsCBMobile2AttVer2", BlipCBMobileReplicationV2, AttVersion2},
+		{"TestAllowedAttachmentsCBMobile3AttVer1", BlipCBMobileReplicationV3, AttVersion1},
+		{"TestAllowedAttachmentsCBMobile3AttVer2", BlipCBMobileReplicationV3, AttVersion2},
+	}
+
+	requireIsAttachmentAllowedTrue := func(t *testing.T, ctx *BlipSyncContext, docID string, meta []AttachmentStorageMeta, activeSubprotocol string) {
+		for _, att := range meta {
+			key := allowedAttachmentKey(docID, att.digest, activeSubprotocol)
+			require.True(t, ctx.isAttachmentAllowed(key))
+		}
+	}
+
+	requireIsAttachmentAllowedFalse := func(t *testing.T, ctx *BlipSyncContext, docID string, meta []AttachmentStorageMeta, activeSubprotocol string) {
+		for _, att := range meta {
+			key := allowedAttachmentKey(docID, att.digest, activeSubprotocol)
+			require.False(t, ctx.isAttachmentAllowed(key))
+		}
+	}
+
+	// Single document associated with multiple attachments of different digests.
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := &BlipSyncContext{}
+			meta := []AttachmentStorageMeta{
+				{digest: "digest1", version: tt.inputAttVersion},
+				{digest: "digest2", version: tt.inputAttVersion},
+			}
+			var docID = "doc1"
+
+			ctx.addAllowedAttachments(docID, meta, tt.inputBlipProtocol)
+			requireIsAttachmentAllowedTrue(t, ctx, docID, meta, tt.inputBlipProtocol)
+
+			ctx.removeAllowedAttachments(docID, meta, tt.inputBlipProtocol)
+			requireIsAttachmentAllowedFalse(t, ctx, docID, meta, tt.inputBlipProtocol)
+		})
+	}
+
+	// Single document associated with multiple attachments of similar digests.
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := &BlipSyncContext{}
+			meta := []AttachmentStorageMeta{
+				{digest: "digest1", version: tt.inputAttVersion},
+				{digest: "digest1", version: tt.inputAttVersion},
+			}
+			var docID = "doc1"
+
+			ctx.addAllowedAttachments(docID, meta, tt.inputBlipProtocol)
+			key := allowedAttachmentKey(docID, meta[0].digest, tt.inputBlipProtocol)
+			require.True(t, ctx.isAttachmentAllowed(key))
+
+			ctx.removeAllowedAttachments(docID, meta, tt.inputBlipProtocol)
+			require.False(t, ctx.isAttachmentAllowed(key))
+		})
+	}
+
+	// Multiple documents associated with multiple attachments of different digests.
```

Comment:
You don't seem to have the simple case of two document with different attachments - here you're attaching the same two attachments to different documents.  Might be nice to have the basic case as well, to make sure it doesn't get broken.